### ISSUE TITLE
Algorithms related papers

### ISSUE THREAD

**Issue by A-lucky-boy** [sasam2/nesting2#2]
Hello author, your program is very good, may I ask when you write the algorithm, have a reference paper? I'd like to take a closer look at your algorithm

**Comment by sasam2** [comment]
Hello, you can check the paper here:
[https://www.dropbox.com/scl/fi/1s0o4ju5yslga42celg8j/sofia_io_2013.pdf?rlkey=9k2tl1u3vx83wl1gontbdv5an&e=1&dl=0](https://www.dropbox.com/scl/fi/1s0o4ju5yslga42celg8j/sofia_io_2013.pdf?rlkey=9k2tl1u3vx83wl1gontbdv5an&e=1&dl=0)
Please download it, as I may delete it in the future.

**Comment by A-lucky-boy** [comment]
> 您好，您可以在这里查看论文： https: [//www.dropbox.com/scl/fi/1s0o4ju5yslga42celg8j/sofia_io_2013.pdf](https://www.dropbox.com/scl/fi/1s0o4ju5yslga42celg8j/sofia_io_2013.pdf?rlkey=9k2tl1u3vx83wl1gontbdv5an&e=1&dl=0) ?rlkey=9k2tl1u3vx83wl1gontbdv5an&e=1&dl=0 请下载它，因为我将来可能会删除它。

Thank you very much for your reply and project, I have downloaded it, wish you a happy life

**Comment by A-lucky-boy** [comment]
> 您好，您可以在这里查看论文： [https://www.dropbox.com/scl/fi/1s0o4ju5yslga42celg8j/sofia_io_2013.pdf?](https://www.dropbox.com/scl/fi/1s0o4ju5yslga42celg8j/sofia_io_2013.pdf?rlkey=9k2tl1u3vx83wl1gontbdv5an&e=1&dl=0) rlkey=9k2tl1u3vx83wl1gontbdv5an&e=1&dl=0 请下载，因为我可能会在将来删除它。

Hello author, I have another question, can you handle parts polygons with holes in your project?

**Comment by sasam2** [comment]
Yes, it can handle polygons with holes (it works by decomposing them into simpler convex polygons).
There's only one thing it does not handle which are exact fits (when a polygon fits exactly in one space, such as the piece of a puzzle). This is because an exact fit is represented by a **point** in the NFP's image. As we are using a raster representation, the point will not be visible in the NFP's image because it has no dimension and this placement will be missed. For further information please look at the last paragraph of page 11 of the article and image 11a.

**Comment by A-lucky-boy** [comment]
> Yes, it can handle polygons with holes (it works by decomposing them into simpler convex polygons). There's only one thing it does not handle which are exact fits (when a polygon fits exactly in one space, such as the piece of a puzzle). This is because an exact fit is represented by a **point** in the NFP's image. As we are using a raster representation, the point will not be visible in the NFP's image because it has no dimension and this placement will be missed. For further information please look at the last paragraph of page 11 of the article and image 11a.

Hello, during this time, I have studied your paper and code more carefully and find the project increasingly interesting. However, I encountered an error when attempting to nest polygons with holes. I used the same method for inputting the polygons with holes as I did for the sheet material, but the program still throws an error. Could you please provide an input template for parts with holes?

**Comment by sasam2** [comment]
I've just checked and you are right: this solution does not support polygons with holes. I tried adding the following polygon (simulating a polygon with a hole) to the existing shapes.txt problem:
```
PIECE 5
QUANTITY
2
NUMBER OF VERTICES
13
VERTICES (X,Y)
 0   0
 0  -1
-2  -1
-2   2
 2   2
 2  -1
 0  -1
 0   0
 1   0
 1   1
-1   1
-1   0
-2  -1
```
and an exception is thrown during the decompositon into convex polygons, which means that this shape is not accepted. :/

**Comment by A-lucky-boy** [comment]
> I've just checked and you are right: this solution does not support polygons with holes. I tried adding the following polygon (simulating a polygon with a hole) to the existing shapes.txt problem:
> 
> ```
> PIECE 5
> QUANTITY
> 2
> NUMBER OF VERTICES
> 13
> VERTICES (X,Y)
>  0   0
>  0  -1
> -2  -1
> -2   2
>  2   2
>  2  -1
>  0  -1
>  0   0
>  1   0
>  1   1
> -1   1
> -1   0
> -2  -1
> ```
> 
> and an exception is thrown during the decompositon into convex polygons, which means that this shape is not accepted. :/

Thank you very much for your reply, I really like your project.

**Comment by A-lucky-boy** [comment]
> I've just checked and you are right: this solution does not support polygons with holes. I tried adding the following polygon (simulating a polygon with a hole) to the existing shapes.txt problem:
> 
> ```
> PIECE 5
> QUANTITY
> 2
> NUMBER OF VERTICES
> 13
> VERTICES (X,Y)
>  0   0
>  0  -1
> -2  -1
> -2   2
>  2   2
>  2  -1
>  0  -1
>  0   0
>  1   0
>  1   1
> -1   1
> -1   0
> -2  -1
> ```
> 
> and an exception is thrown during the decompositon into convex polygons, which means that this shape is not accepted. :/

Hello author, sorry to bother you again, I would like to ask, if I want to arrange samples on two boards at the same time, can I achieve it by modifying the txt file
